### ISSUE TITLE
AWS EKS w/ ENI, prefixDelegation, custom CIDR and bottlerocket running out of IPs

### ISSUE THREAD

**Issue by yrotilio** [cilium/cilium#29634]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### What happened?

On an EKS cluster (1.28) with ENI mode, prefix delegation and custom CIDR (100.64.0.0/16) activated, we have a few nodes on one cluster with many Pods stuck in ContainerCreating status with the following event:
```
Events:
  Type     Reason                  Age                     From     Message
  ----     ------                  ----                    ----     -------
  Warning  FailedCreatePodSandBox  2m23s (x476 over 105m)  kubelet  (combined from similar events): Failed to create pod sandbox: rpc error: code = Unknown desc = failed to setup network for sandbox "c6f7d9434a5b18602b4329dca397d7731b99fccc0e0baacaeaf82a77d9c9439e": plugin type="cilium-cni" failed (add): unable to allocate IP via local cilium agent: [POST /ipam][502] postIpamFailure  No more IPs available
```
Yet there are still quite a few free IPs in the subnet according to AWS Console.
![Capture d'écran 2023-12-05 135016](https://github.com/cilium/cilium/assets/10231065/8670490b-5d46-4ee3-9a25-45e3777f4940)

Subnets are 3*/23 on 3 AZs, which should allow for ~1507 IPs to be allocated.
Cluster shows ~542 Pods on ~13 Nodes, and CiliumNodes CRD seems to show ~787 IPs allocated, which seems high compared to Pod count but shouldn't block IP allocation.
```
$ kubectl get nodes -oname | wc -l
13                              

$ kubectl get pods -oname -A | wc -l
542                              

$ kubectl get ciliumnodes.cilium.io -ojson | jq -r '.items[].status.eni.enis[].addresses' | jq -r '.[]' 2>>/dev/null | wc -l
787
```

We then attempted the following resolutions:
* We've tried updating our configuration with nodeinit.enable=true (from https://github.com/cilium/cilium/issues/22054), but it's not compatible with Bottlerocket (it fails because there is no bash on the Node)
* We've tried to change the pre-allocate parameter (https://docs.cilium.io/en/stable/network/concepts/ipam/eni/#eni-allocation-parameters), but putting it in cni-configuration configmap or any other config does'nt seem to have any impact on CiliumNodes configuration.
* We've tried rolling out cilium-operator, cilium-agent and even nodes, but in the end we still have at least one node with the issue.

Doing that, we realized that PrefixDelegation does'nt seem consistent, for we have some ENIs with secondary IPs only, some with prefixes only, and some with both.
![Capture d'écran 2023-12-05 140052](https://github.com/cilium/cilium/assets/10231065/bdada4dc-0b75-46e3-a149-691cfdcdaaca)

Furthermore, looking at logs, it seems like sometimes the parameter changes during the lifecycle of an instance:
```
level=info msg="Resolving IP deficit of node" available=0 availableForAllocation=0 emptyInterfaceSlots=2 instanceID=i-0d066069bbb5ed99e maxIPsToAllocate=8 name=ip-10-170-17-53.eu-west-3.compute.internal neededIPs=8 remainingInterfaces=2 selectedInterface= selectedPoolID= subsys=ipam used=0
level=info msg="No more IPs available, creating new ENI" addresses=5 instanceID=i-0d066069bbb5ed99e isPrefixDelegated=true name=ip-10-170-17-53.eu-west-3.compute.internal securityGroupIDs="[sg-********************]" subnetID=subnet-******************** subsys=ipam
level=warning msg="Subnet might be out of prefixes, Cilium will not allocate prefixes on this node anymore" addresses=5 instanceID=i-0d066069bbb5ed99e isPrefixDelegated=true name=ip-10-170-17-53.eu-west-3.compute.internal node=ip-10-170-17-53.eu-west-3.compute.internal securityGroupIDs="[sg-********************]" subnetID=subnet-******************** subsys=ipam
level=info msg="Created new ENI" addresses=5 eniID=eni-0ae23800af1cd0d41 instanceID=i-0d066069bbb5ed99e isPrefixDelegated=true name=ip-10-170-17-53.eu-west-3.compute.internal securityGroupIDs="[sg-********************]" subnetID=subnet-******************** subsys=ipam
level=info msg="Attached ENI to instance" addresses=5 attachmentID=eni-attach-0f1bf99b7bc9c0f70 eniID=eni-0ae23800af1cd0d41 index=1 instanceID=i-0d066069bbb5ed99e isPrefixDelegated=true name=ip-10-170-17-53.eu-west-3.compute.internal securityGroupIDs="[sg-********************]" subnetID=subnet-******************** subsys=ipam
level=info msg="Resolving IP deficit of node" available=5 availableForAllocation=0 emptyInterfaceSlots=1 instanceID=i-0d066069bbb5ed99e maxIPsToAllocate=4 name=ip-10-170-17-53.eu-west-3.compute.internal neededIPs=3 remainingInterfaces=1 selectedInterface= selectedPoolID= subsys=ipam used=0
level=info msg="No more IPs available, creating new ENI" addresses=4 instanceID=i-0d066069bbb5ed99e isPrefixDelegated=false name=ip-10-170-17-53.eu-west-3.compute.internal securityGroupIDs="[sg-********************]" subnetID=subnet-******************** subsys=ipam
level=info msg="Created new ENI" addresses=4 eniID=eni-025643667709b0534 instanceID=i-0d066069bbb5ed99e isPrefixDelegated=false name=ip-10-170-17-53.eu-west-3.compute.internal securityGroupIDs="[sg-********************]" subnetID=subnet-******************** subsys=ipam
level=info msg="Attached ENI to instance" addresses=4 attachmentID=eni-attach-0f6b2bf8e0ee1be5b eniID=eni-025643667709b0534 index=2 instanceID=i-0d066069bbb5ed99e isPrefixDelegated=false name=ip-10-170-17-53.eu-west-3.compute.internal securityGroupIDs="[sg-********************]" subnetID=subnet-******************** subsys=ipam                          
```

As of now, we have no idea if this behavior on Prefix Delegation is in any way related to our IP availability issue, but it seems sufficiently important to be mentioned. Do not hesitate to tell me if I need to create a separate issue.

### Cilium Version

Client: 1.14.1 c191ef6f 2023-08-10T18:54:57+02:00 go version go1.20.7 linux/amd64
Daemon: 1.14.1 c191ef6f 2023-08-10T18:54:57+02:00 go version go1.20.7 linux/amd64   

### Kernel Version

Linux version 6.1.59

### Kubernetes Version

Client Version: v1.28.4
Kustomize Version: v5.0.4-0.20230601165947-6ce0bf390ce3
Server Version: v1.28.3-eks-4f4795d     

### Sysdump

sysdump is actually too big to be uploaded on github so I had to split it.
[cilium-sysdump-20231205-141320-1.zip](https://github.com/cilium/cilium/files/13561471/cilium-sysdump-20231205-141320-1.zip)
[cilium-sysdump-20231205-141320-2.zip](https://github.com/cilium/cilium/files/13561476/cilium-sysdump-20231205-141320-2.zip)

### Relevant log output

```shell
apiVersion: cilium.io/v2
kind: CiliumNode
metadata:
  creationTimestamp: "2023-12-04T15:28:22Z"
  generation: 19
  labels:
    beta.kubernetes.io/arch: amd64
    beta.kubernetes.io/instance-type: r6i.2xlarge
    beta.kubernetes.io/os: linux
    failure-domain.beta.kubernetes.io/region: eu-west-3
    failure-domain.beta.kubernetes.io/zone: eu-west-3c
    k8s.io/cloud-provider-aws: 907410ba2d72b08f6b0105bead9472df
    karpenter.k8s.aws/instance-category: r
    karpenter.k8s.aws/instance-cpu: "8"
    karpenter.k8s.aws/instance-encryption-in-transit-supported: "true"
    karpenter.k8s.aws/instance-family: r6i
    karpenter.k8s.aws/instance-generation: "6"
    karpenter.k8s.aws/instance-hypervisor: nitro
    karpenter.k8s.aws/instance-memory: "65536"
    karpenter.k8s.aws/instance-network-bandwidth: "3125"
    karpenter.k8s.aws/instance-pods: "110"
    karpenter.k8s.aws/instance-size: 2xlarge
    karpenter.sh/capacity-type: on-demand
    karpenter.sh/initialized: "true"
    karpenter.sh/provisioner-name: projet-g1
    karpenter.sh/registered: "true"
    kubernetes.io/arch: amd64
    kubernetes.io/hostname: ip-10-170-17-19.eu-west-3.compute.internal
    kubernetes.io/os: linux
    node.kubernetes.io/instance-type: r6i.2xlarge
    topology.ebs.csi.aws.com/zone: eu-west-3c
    topology.kubernetes.io/region: eu-west-3
    topology.kubernetes.io/zone: eu-west-3c
  name: ip-10-170-17-19.eu-west-3.compute.internal
  ownerReferences:
  - apiVersion: v1
    kind: Node
    name: ip-10-170-17-19.eu-west-3.compute.internal
    uid: e803f0e1-ee26-411c-b25c-12176e0fa9c0
  resourceVersion: "11034099"
  uid: 4fee97ec-cf02-4392-bb84-2c06b515c05b
spec:
  addresses:
  - ip: 10.170.17.19
    type: InternalIP
  - ip: 100.64.11.110
    type: CiliumInternalIP
  alibaba-cloud: {}
  azure: {}
  encryption: {}
  eni:
    availability-zone: eu-west-3c
    disable-prefix-delegation: false
    first-interface-index: 1
    instance-type: r6i.2xlarge
    node-subnet-id: subnet-************
    subnet-tags:
      subnet-type: pods
    use-primary-address: false
    vpc-id: vpc-**************
  health:
    ipv4: 100.64.11.252
  ingress: {}
  instance-id: i-06a4e5af6e589b2f5
  ipam:
    podCIDRs:
    - 10.19.0.0/16
    pool:
      100.64.10.4:
        resource: eni-03117699d7576cebd
      100.64.10.12:
        resource: eni-0adfe34d0abc9f50a
      100.64.10.52:
        resource: eni-09fea04c8ecd8e38b
      100.64.10.58:
        resource: eni-09fea04c8ecd8e38b
      100.64.10.61:
        resource: eni-09fea04c8ecd8e38b
      100.64.10.83:
        resource: eni-0adfe34d0abc9f50a
      100.64.10.162:
        resource: eni-03117699d7576cebd
      100.64.10.182:
        resource: eni-0adfe34d0abc9f50a
      100.64.10.185:
        resource: eni-0adfe34d0abc9f50a
      100.64.10.186:
        resource: eni-0adfe34d0abc9f50a
      100.64.10.192:
        resource: eni-03117699d7576cebd
      100.64.10.193:
        resource: eni-03117699d7576cebd
      100.64.10.194:
        resource: eni-03117699d7576cebd
      100.64.10.195:
        resource: eni-03117699d7576cebd
      100.64.10.196:
        resource: eni-03117699d7576cebd
      100.64.10.197:
        resource: eni-03117699d7576cebd
      100.64.10.198:
        resource: eni-03117699d7576cebd
      100.64.10.199:
        resource: eni-03117699d7576cebd
      100.64.10.200:
        resource: eni-03117699d7576cebd
      100.64.10.201:
        resource: eni-03117699d7576cebd
      100.64.10.202:
        resource: eni-03117699d7576cebd
      100.64.10.203:
        resource: eni-03117699d7576cebd
      100.64.10.204:
        resource: eni-03117699d7576cebd
      100.64.10.205:
        resource: eni-03117699d7576cebd
      100.64.10.206:
        resource: eni-03117699d7576cebd
      100.64.10.207:
        resource: eni-03117699d7576cebd
      100.64.10.212:
        resource: eni-0adfe34d0abc9f50a
      100.64.10.234:
        resource: eni-0adfe34d0abc9f50a
      100.64.10.242:
        resource: eni-0adfe34d0abc9f50a
      100.64.10.245:
        resource: eni-0adfe34d0abc9f50a
      100.64.11.38:
        resource: eni-0adfe34d0abc9f50a
      100.64.11.43:
        resource: eni-0adfe34d0abc9f50a
      100.64.11.76:
        resource: eni-0adfe34d0abc9f50a
      100.64.11.96:
        resource: eni-0adfe34d0abc9f50a
      100.64.11.97:
        resource: eni-0adfe34d0abc9f50a
      100.64.11.98:
        resource: eni-0adfe34d0abc9f50a
      100.64.11.99:
        resource: eni-0adfe34d0abc9f50a
      100.64.11.100:
        resource: eni-0adfe34d0abc9f50a
      100.64.11.101:
        resource: eni-0adfe34d0abc9f50a
      100.64.11.102:
        resource: eni-0adfe34d0abc9f50a
      100.64.11.103:
        resource: eni-0adfe34d0abc9f50a
      100.64.11.104:
        resource: eni-0adfe34d0abc9f50a
      100.64.11.105:
        resource: eni-0adfe34d0abc9f50a
      100.64.11.106:
        resource: eni-0adfe34d0abc9f50a
      100.64.11.107:
        resource: eni-0adfe34d0abc9f50a
      100.64.11.108:
        resource: eni-0adfe34d0abc9f50a
      100.64.11.109:
        resource: eni-0adfe34d0abc9f50a
      100.64.11.110:
        resource: eni-0adfe34d0abc9f50a
      100.64.11.111:
        resource: eni-0adfe34d0abc9f50a
      100.64.11.202:
        resource: eni-0adfe34d0abc9f50a
      100.64.11.252:
        resource: eni-03117699d7576cebd
    pools: {}
    pre-allocate: 8
status:
  alibaba-cloud: {}
  azure: {}
  eni:
    enis:
      eni-09fea04c8ecd8e38b:
        addresses:
        - 100.64.10.52
        - 100.64.10.58
        - 100.64.10.61
        description: Cilium-CNI (i-06a4e5af6e589b2f5)
        id: eni-09fea04c8ecd8e38b
        ip: 100.64.11.2
        mac: 0e:e5:c4:9e:d2:a8
        number: 3
        security-groups:
        - sg-********************
        subnet:
          cidr: 100.64.10.0/23
          id: subnet-***************
        tags:
          io.cilium/cilium-managed: "true"
          io.cilium/cluster-name: dok-retnp-dax01
          kubernetes.io/cluster/dok-retnp-dax01: owned
        vpc:
          cidrs:
          - 10.170.16.0/20
          - 100.64.0.0/16
          id: vpc-*****************
          primary-cidr: 10.145.96.0/19
      eni-0580732b19cf203e8:
        id: eni-0580732b19cf203e8
        ip: 10.170.17.19
        mac: 0e:75:f5:94:20:e2
        security-groups:
        - sg-******************
        subnet:
          cidr: 10.170.17.0/25
          id: subnet-****************
        tags:
          kubernetes.io/cluster/dok-retnp-dax01: owned
        vpc:
          cidrs:
          - 10.170.16.0/20
          - 100.64.0.0/16
          id: vpc-06f84678df3ad4efc
          primary-cidr: 10.145.96.0/19
      eni-03117699d7576cebd:
        addresses:
        - 100.64.10.162
        - 100.64.10.4
        - 100.64.11.252
        - 100.64.10.192
        - 100.64.10.193
        - 100.64.10.194
        - 100.64.10.195
        - 100.64.10.196
        - 100.64.10.197
        - 100.64.10.198
        - 100.64.10.199
        - 100.64.10.200
        - 100.64.10.201
        - 100.64.10.202
        - 100.64.10.203
        - 100.64.10.204
        - 100.64.10.205
        - 100.64.10.206
        - 100.64.10.207
        description: Cilium-CNI (i-06a4e5af6e589b2f5)
        id: eni-03117699d7576cebd
        ip: 100.64.11.192
        mac: 0e:29:d0:09:17:20
        number: 2
        prefixes:
        - 100.64.10.192/28
        security-groups:
        - sg-*************
        subnet:
          cidr: 100.64.10.0/23
          id: subnet-****************
        tags:
          io.cilium/cilium-managed: "true"
          io.cilium/cluster-name: dok-retnp-dax01
          kubernetes.io/cluster/dok-retnp-dax01: owned
        vpc:
          cidrs:
          - 10.170.16.0/20
          - 100.64.0.0/16
          id: vpc-***************
          primary-cidr: 10.145.96.0/19
      eni-0adfe34d0abc9f50a:
        addresses:
        - 100.64.11.38
        - 100.64.10.234
        - 100.64.11.202
        - 100.64.11.43
        - 100.64.10.12
        - 100.64.11.76
        - 100.64.10.242
        - 100.64.10.83
        - 100.64.10.212
        - 100.64.10.245
        - 100.64.10.182
        - 100.64.10.185
        - 100.64.10.186
        - 100.64.11.96
        - 100.64.11.97
        - 100.64.11.98
        - 100.64.11.99
        - 100.64.11.100
        - 100.64.11.101
        - 100.64.11.102
        - 100.64.11.103
        - 100.64.11.104
        - 100.64.11.105
        - 100.64.11.106
        - 100.64.11.107
        - 100.64.11.108
        - 100.64.11.109
        - 100.64.11.110
        - 100.64.11.111
        description: Cilium-CNI (i-06a4e5af6e589b2f5)
        id: eni-0adfe34d0abc9f50a
        ip: 100.64.10.16
        mac: 0e:df:0b:5d:d7:3a
        number: 1
        prefixes:
        - 100.64.11.96/28
        security-groups:
        - sg-***************
        subnet:
          cidr: 100.64.10.0/23
          id: subnet-*****************
        tags:
          io.cilium/cilium-managed: "true"
          io.cilium/cluster-name: dok-retnp-dax01
          kubernetes.io/cluster/dok-retnp-dax01: owned
        vpc:
          cidrs:
          - 10.170.16.0/20
          - 100.64.0.0/16
          id: vpc-**************
          primary-cidr: 10.145.96.0/19
  ipam:
    operator-status: {}
    used:
      100.64.10.4:
        owner: ns/***********************************
        resource: eni-03117699d7576cebd
      100.64.10.12:
        owner: ns/*********************************** [restored]
        resource: eni-0adfe34d0abc9f50a
      100.64.10.52:
        owner: ns/*********************************** [restored]
        resource: eni-09fea04c8ecd8e38b
      100.64.10.58:
        owner: ns/*********************************** [restored]
        resource: eni-09fea04c8ecd8e38b
      100.64.10.61:
        owner: ns/*********************************** [restored]
        resource: eni-09fea04c8ecd8e38b
      100.64.10.83:
        owner: ns/*********************************** [restored]
        resource: eni-0adfe34d0abc9f50a
      100.64.10.162:
        owner: ns/*********************************** [restored]
        resource: eni-03117699d7576cebd
      100.64.10.182:
        owner: ns/*********************************** [restored]
        resource: eni-0adfe34d0abc9f50a
      100.64.10.185:
        owner: ns/*********************************** [restored]
        resource: eni-0adfe34d0abc9f50a
      100.64.10.186:
        owner: ns/*********************************** [restored]
        resource: eni-0adfe34d0abc9f50a
      100.64.10.192:
        owner: ns/*********************************** [restored]
        resource: eni-03117699d7576cebd
      100.64.10.193:
        owner: ns/*********************************** [restored]
        resource: eni-03117699d7576cebd
      100.64.10.194:
        owner: ns/*********************************** [restored]
        resource: eni-03117699d7576cebd
      100.64.10.195:
        owner: ns/*********************************** [restored]
        resource: eni-03117699d7576cebd
      100.64.10.196:
        owner: ns/*********************************** [restored]
        resource: eni-03117699d7576cebd
      100.64.10.197:
        owner: ns/*********************************** [restored]
        resource: eni-03117699d7576cebd
      100.64.10.198:
        owner: ns/*********************************** [restored]
        resource: eni-03117699d7576cebd
      100.64.10.199:
        owner: ns/*********************************** [restored]
        resource: eni-03117699d7576cebd
      100.64.10.200:
        owner: ns/*********************************** [restored]
        resource: eni-03117699d7576cebd
      100.64.10.201:
        owner: ns/*********************************** [restored]
        resource: eni-03117699d7576cebd
      100.64.10.202:
        owner: ns/*********************************** [restored]
        resource: eni-03117699d7576cebd
      100.64.10.203:
        owner: ns/*********************************** [restored]
        resource: eni-03117699d7576cebd
      100.64.10.204:
        owner: ns/*********************************** [restored]
        resource: eni-03117699d7576cebd
      100.64.10.205:
        owner: ns/*********************************** [restored]
        resource: eni-03117699d7576cebd
      100.64.10.206:
        owner: ns/*********************************** [restored]
        resource: eni-03117699d7576cebd
      100.64.10.207:
        owner: ns/*********************************** [restored]
        resource: eni-03117699d7576cebd
      100.64.10.212:
        owner: ns/*********************************** [restored]
        resource: eni-0adfe34d0abc9f50a
      100.64.10.234:
        owner: ns/*********************************** [restored]
        resource: eni-0adfe34d0abc9f50a
      100.64.10.242:
        owner: ns/*********************************** [restored]
        resource: eni-0adfe34d0abc9f50a
      100.64.10.245:
        owner: ns/*********************************** [restored]
        resource: eni-0adfe34d0abc9f50a
      100.64.11.38:
        owner: ns/***********************************
          [restored]
        resource: eni-0adfe34d0abc9f50a
      100.64.11.43:
        owner: ns/*********************************** [restored]
        resource: eni-0adfe34d0abc9f50a
      100.64.11.76:
        owner: ns/***********************************x [restored]
        resource: eni-0adfe34d0abc9f50a
      100.64.11.96:
        owner: ns/*********************************** [restored]
        resource: eni-0adfe34d0abc9f50a
      100.64.11.97:
        owner: ns/***********************************
        resource: eni-0adfe34d0abc9f50a
      100.64.11.98:
        owner: ns/*********************************** [restored]
        resource: eni-0adfe34d0abc9f50a
      100.64.11.99:
        owner: ns/*********************************** [restored]
        resource: eni-0adfe34d0abc9f50a
      100.64.11.100:
        owner: ns/*********************************** [restored]
        resource: eni-0adfe34d0abc9f50a
      100.64.11.101:
        owner: ns/*********************************** [restored]
        resource: eni-0adfe34d0abc9f50a
      100.64.11.102:
        owner: ns/*********************************** [restored]
        resource: eni-0adfe34d0abc9f50a
      100.64.11.103:
        owner: ns/*********************************** [restored]
        resource: eni-0adfe34d0abc9f50a
      100.64.11.104:
        owner: kube-system/node-local-dns-9nnsb [restored]
        resource: eni-0adfe34d0abc9f50a
      100.64.11.105:
        owner: ns/*********************************** [restored]
        resource: eni-0adfe34d0abc9f50a
      100.64.11.106:
        owner: ns/*********************************** [restored]
        resource: eni-0adfe34d0abc9f50a
      100.64.11.107:
        owner: ns/*********************************** [restored]
        resource: eni-0adfe34d0abc9f50a
      100.64.11.108:
        owner: ns/*********************************** [restored]
        resource: eni-0adfe34d0abc9f50a
      100.64.11.109:
        owner: ns/*********************************** [restored]
        resource: eni-0adfe34d0abc9f50a
      100.64.11.110:
        owner: router
        resource: eni-0adfe34d0abc9f50a
      100.64.11.111:
        owner: kube-infra/efs-csi-node-xbrxz [restored]
        resource: eni-0adfe34d0abc9f50a
      100.64.11.202:
        owner: ns/*********************************** [restored]
        resource: eni-0adfe34d0abc9f50a
      100.64.11.252:
        owner: health
        resource: eni-03117699d7576cebd
```

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by yrotilio** [comment]
Hi @ti-mo, thanks for taking the time to answer my humble issue !

I'm well aware of that limit, especially considering that I use an extended CIDR (100.64.0.0/XX) that consume a whole ENI just for routing to work. 

A way to workaround that limit on EC2 Instances is to use Prefix Delegation, which allows to assign /28 prefixes instead of IPs to ENIs, actually multiplying the number of IPs per ENI by 12. That feature is well described and documented on AWS and on AWS VPC CNI (which I have used for a few years) but does not appear in Cilium Documentation except in helm chart [values](https://docs.cilium.io/en/stable/helm-reference/#id1).
Considering that [comment](https://github.com/cilium/cilium/blob/c0dfeb0f8460f44c3983d5bfa15b6afcdc5baf51/operator/option/config.go#L384) in Cilium code though, I understand it's supposed to allow for the same behavior. That is confirmed by the fact that right now, I have one r6i.2xlarge node with 101 pods Running.

Sorry if I wasn't clear enough in my issue, my problem are that:
- Pods are stuck on ContainerCreating Status because Cilium does not allocate an IP or a Prefix to a Node although there are IPs available in the Subnet.
- Cilium does not seem to have a consistent behavior toward Prefix Delegation allocation.

**Comment by AlenversFr** [comment]
We are still stuck with this issu.

As for now if we have 500 pods running in a cluster we have to allocate /21 CIDR (2048 IPs)

**Comment by slenky** [comment]
Cilium 1.14.6, we started to face the same issue after upgrading from EKS 1.27 to EKS 1.29.

**Comment by hemanthmalla** [comment]
@yrotilio you may be running into IP space fragmentation in your subnets. There maybe enough /32 IPs available but not enough contiguous /28 blocks available. 

Once the operator sees a mix of prefixes and secondary IPs available, operator will not attempt to allocate prefixes on that node anymore. Operator will retry allocating prefixes for every new node that comes up.

**Comment by yrotilio** [comment]
Hi @hemanthmalla, thanks for taking the time to look into my issue.

I do confirm that the main problem seems to be IP fragmentation in our subnets, and that there is not enough contiguous IP space to allocate prefixes. What I don't understand is why cilium doesn't fallback as it should : allocating /32 IP adresses if there is no /28 continous block available.

But I've just realised 1.15.2 has been released with a specific improvement on that part thanks to you, so i'm gonna try it out and see if it changes anything.

I've also had a meeting with AWS CNI guys, and they suggested using CIDR prefix reservation to seperate /32 and /28 allocation to limit fragmentation. I'm not sure how cilium will react to that, do you have any insight on that end ?

**Comment by jschwartzy** [comment]
> I've also had a meeting with AWS CNI guys, and they suggested using CIDR prefix reservation to seperate /32 and /28 allocation to limit fragmentation. I'm not sure how cilium will react to that, do you have any insight on that end ?

We've encountered the fragmentation issue and setup Subnet CIDR reservations to block pools for the prefixes and left the rest of the CIDR unreserved for ENI/other assignments. It works really well - afaik Cilium Operator is just making the API call to get a prefix and it's up to AWS to provide the allocation, which takes the reservation rules into account. 

Two things to look out for:
1. Once you reserve the IPs, the Subnet IP availability count will immediately be deducted, so you will need to figure out another solution for monitoring IP availability 
2 You will need to rotate your nodes so they pull from the correct pools since reservations don't act on IPs that are already assigned.

**Comment by christarazi** [comment]
> But I've just realised 1.15.2 has been released with a specific improvement on that part thanks to you, so i'm gonna try it out and see if it changes anything.

@yrotilio Has the issue been resolved for you in 1.15.2?
